shaping/wrapping.go

```diff
@@ -362,6 +363,200 @@ func (r *runMapper) mapRun(runIdx int, run Output) {
+
+// Peek implements [RunIterator.Peek].
+func (r *runSlice) Peek() (int, Output, bool) {
+	if r.idx < len(r.runs) {
+		next := r.runs[r.idx]
+		return r.idx, next, true
+	}
+	return r.idx, Output{}, false
+}
+
+// Save implements [RunIterator.Save].
+func (r *runSlice) Save() {
+	r.savedIdx = r.idx
+}
+
+// Restore implements [RunIterator.Restore].
+func (r *runSlice) Restore() {
+	r.idx = r.savedIdx
+}
+
+// WrapBuffer provides reusable buffers for line wrapping. When using a
+// WrapBuffer, returned line wrapping results will use memory stored within
+// the buffer. This means that the same buffer cannot be reused for another
+// wrapping operation while the wrapped lines are still in use (unless they
+// are deeply copied). If necessary, using a multiple WrapBuffers can work
+// around this restriction.
+type WrapBuffer struct {
+	// paragraph is a buffer holding paragraph allocated (primarily) from subregions
+	// of the line field.
+	paragraph []Line
+	// line is a large buffer of Outputs that is used to build lines.
+	line []Output
+	// lineUsed is the index of the first unused element in line.
+	lineUsed int
+	// lineExhausted indicates whether the previous shaping used all of line.
+	lineExhausted bool
+	// alt is a smaller temporary buffer that holds candidate lines while
+	// they are being built.
+	alt []Output
+	// altAdvance is the sum of the advances of each run in alt.
+	altAdvance fixed.Int26_6
+	// best is a slice holding the best known line. When possible, it
+	// is a subslice of line, but if line runs out of capacity it will
+	// be heap allocated.
+	best []Output
+	// bestInLine tracks whether the current best is allocated from within line.
+	bestInLine bool
+}
+
+// NewWrapBuffer returns a WrapBuffer with some pre-allocated storage for
+// small-sized texts.
+func NewWrapBuffer() *WrapBuffer {
+	return &WrapBuffer{
+		paragraph: make([]Line, 0, 10),
+		line:      make([]Output, 0, 100),
+		alt:       make([]Output, 0, 10),
+	}
+}
+
+func (w *WrapBuffer) reset() {
+	w.paragraph = w.paragraph[:0]
+	w.alt = w.alt[:0]
+	w.altAdvance = 0
+	w.line = w.line[:0]
+	w.lineUsed = 0
+	w.best = nil
+	w.bestInLine = false
+	if w.lineExhausted {
+		w.lineExhausted = false
+		// Trigger the go slice growth heuristic by appending an element to
+		// the capacity.
+		w.line = append(w.line[:cap(w.line)], Output{})[:0]
+	}
+}
+
+func (w *WrapBuffer) stats() string {
+	return fmt.Sprintf("paragraph: %d(%d), line: %d(%d), used: %d, exhausted: %v, alt: %d(%d)", len(w.paragraph), cap(w.paragraph), len(w.line), cap(w.line), w.lineUsed, w.lineExhausted, len(w.alt), cap(w.alt))
+}
+
+// singleRunParagraph is an optimized helper for quickly constructing
+// a []Line containing only a single run.
+func (w *WrapBuffer) singleRunParagraph(run Output) []Line {
+	w.paragraph = w.paragraph[:0]
+	s := w.line[w.lineUsed : w.lineUsed+1]
+	s[0] = run
+	w.paragraphAppend(s)
+	return w.finalParagraph()
+}
+
+func (w *WrapBuffer) paragraphAppend(line []Output) {
```

Comment:
Do we really need a method for a "simple" append ?

Reply:
Strictly speaking no, but:

- it will be inlined
- not using a method would make this the only operation relying on directly accessing the buffer's fields. Since the rest of it has a coherent API, I thought this should match.
- we may choose to implement this more cleverly in the future with something like an arraylist, in which case it stops being a trivial append

Reply:
Alright, it makes sense !